sql/src/main/codegen/includes/insert.ftl

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+SqlNode DruidSqlInsert() :
+{
+ SqlNode insertNode;
+ SqlNode partitionBy = null;
+ SqlNodeList clusterBy = null;
+}
+{
+    insertNode = SqlInsert()
+    [
+      <PARTITION> <BY>
+      partitionBy = StringLiteral()
```

Comment:
This does look nicer. Is the `TimeUnit or ALL` abstraction defined somewhere in Druid's codebase which can be reused? Or should I create a new SQL-specific class that encapsulates this?

Reply:
"TimeUnit or ALL" isn't defined, although maybe we could use PeriodGranularity as the class, and have the parser generate a PeriodGranularity directly?

Reply:
Per Gian (and Hive: https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DML#LanguageManualDML-InsertingvaluesintotablesfromSQL) the arguments should be expressions, not literals. Druid may have a very limited set of expressions (such as those Gian showed), but they are still expressions.

Reply:
Also, though I personally cringe at the idea, seems some folks like using the archaic`GROUP BY 3, 2, 5` form in `SELECT` statements. Should we allow ordinals here also? I don't like encouraging their use, but if folks already use them...

Reply:
> Also, though I personally cringe at the idea, seems some folks like using the archaic GROUP BY 3, 2, 5 form in SELECT statements. Should we allow ordinals here also?

I like that GROUP BY syntax because it saves typing 🙂. However, I don't think it's necessary to implement it for PARTITION BY. It'll add a bunch of complexity to the implementation and I don't think it's necessary to support it for every clause.

Reply:
As to what goes after PARTITION BY, I'm torn as to what's best. Context: today, partitioning must either be a floor of the primary timestamp `__time`, or "all" which means everything goes in a single partition. So I think there are three options.

1. `PARTITION BY 'day'` or `PARTITION BY 'all'`. This is what's in the PR now.
2. `PARTITION BY DAY` or `PARTITION BY ALL`. I suggested this upthread. DAY and ALL are keywords, not expressions.
3. `PARTITION BY FLOOR(__TIME TO DAY)`. I suggested this in #11929. In this case the partition-by is an expression. (I can't make up my mind, I guess, so I suggested two different things.) I didn't have a suggestion for how to say "partition by all", but IMO it should be explicit and required, so people don't accidentally partition by all when they didn't mean to.

(3) is closest to what other databases do, but those other databases generally allow partitioning on things other than time-floor. For us, it'd have to be an error to specify anything other than `FLOOR(__time TO XXX)`. That's what makes me think (2) is good: why make people specify the whole expression if it can't be changed? The obvious answer is that making things explicit is good and also that lining things up with other databases is good. But, still…

Anyway. Curious what other people think.

Reply:
I initially suggested (3) but then moved back to camp (1) / (2) because we can only partition on time.  `PARTITIONED BY DAY/ALL`  is a much easier syntax to write. It's also probably easy to give typeahead hints with (2) (console showing a list of all possible granularities as you type PARTITIONED BY).  

I, for one, don't have a strong opinion between (1) and (2).

Reply:
If I had to pick, 2 seems like a cleaner option. 
While 1 is implemented right now, quoting and adding a literal doesn't seem pretty, given that we have a fixed range of what can go in front of a `PARTITIONED BY`.
3 seems a bit more verbose, and to me, it implies that a) I can use a column other than `__time` b) I can use an expression other than `FLOOR`. However, I am unfamiliar with the other databases which implement the syntax in this manner and therefore of the conventions when it comes to `PARTITIONED BY`, hence I am using this convention.

Reply:
One more question regarding the query granularities: `TimeUnit()` in Calcite is defined to have the following units (from MILLISECOND to MILLENNIUM). 
```
    <MILLISECOND> { return TimeUnit.MILLISECOND; }
|   <SECOND> { return TimeUnit.SECOND; }
|   <MINUTE> { return TimeUnit.MINUTE; }
|   <HOUR> { return TimeUnit.HOUR; }
|   <DAY> { return TimeUnit.DAY; }
|   <DOW> { return TimeUnit.DOW; }
|   <DOY> { return TimeUnit.DOY; }
|   <ISODOW> { return TimeUnit.ISODOW; }
|   <ISOYEAR> { return TimeUnit.ISOYEAR; }
|   <WEEK> { return TimeUnit.WEEK; }
|   <MONTH> { return TimeUnit.MONTH; }
|   <QUARTER> { return TimeUnit.QUARTER; }
|   <YEAR> { return TimeUnit.YEAR; }
|   <EPOCH> { return TimeUnit.EPOCH; }
|   <DECADE> { return TimeUnit.DECADE; }
|   <CENTURY> { return TimeUnit.CENTURY; }
|   <MILLENNIUM> { return TimeUnit.MILLENNIUM; }
``` 
Whereas in Druid's TimeUnitRange-Period map, the granularities are defined as 
```
  private static final Map<TimeUnitRange, Period> PERIOD_MAP = ImmutableMap.<TimeUnitRange, Period>builder()
      .put(TimeUnitRange.SECOND, Period.seconds(1))
      .put(TimeUnitRange.MINUTE, Period.minutes(1))
      .put(TimeUnitRange.HOUR, Period.hours(1))
      .put(TimeUnitRange.DAY, Period.days(1))
      .put(TimeUnitRange.WEEK, Period.weeks(1))
      .put(TimeUnitRange.MONTH, Period.months(1))
      .put(TimeUnitRange.QUARTER, Period.months(3))
      .put(TimeUnitRange.YEAR, Period.years(1))
      .build();
```
i.e. there's a mismatch as to what Calcite's SQL can accept as TimeUnit, and what we can convert using TimeUnits class. 
For the ingestion's `PARTITION BY` what range of time units from the above should be accepted by the SQL?